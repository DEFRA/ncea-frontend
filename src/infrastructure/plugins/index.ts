export { customHapiPino } from './logger';
export { customHapiRoutes } from './router';
export { customHapiViews } from './views';

