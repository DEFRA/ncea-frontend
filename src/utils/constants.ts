import { IFilterOptions } from '../interfaces/searchPayload.interface';
import { TabOptions } from '../interfaces/detailsTab.interface';

export const webRoutePaths = {
  home: '/',
  results: '/search',
  guidedDateSearch: '/date-search',
  geographySearch: '/coordinate-search',
  intermediate: '/intermediate',
  getMapResults: '/map-results',
  getMapFilters: '/map-filters',
  filterResourceType: '/resource-type-filter',
  filterStudyPeriod: '/study-period-filter',
  sortResults: '/sort-results',
  help: '/help',
  accessibilityStatement: '/accessibility-statement',
  termsAndConditions: '/terms-conditions',
  privacyPolicy: '/privacy-policy',
  cookiePolicy: '/cookie-policy',
};

export const elasticSearchAPIPaths = {
  searchPath: '_search',
  countPath: '_count',
};

export const formKeys = {
  dateQuestionnaire: 'date-questionnaire',
};

export const formIds = {
  quickSearchFID: 'keyword',
  dataQuestionnaireFID: 'date',
  geographyQuestionnaireFID: 'extent',
};

export const showMoreText: string = 'Show more';
export const showLessText: string = 'Show less';
export const maxWords: number = 100;
export const geoNetworkIndex: string = 'gn-records';
export const mapResultMaxCount: number = 10000;
export const requiredFieldsForMap: string[] = [
  'resourceTitleObject',
  'contactForResource',
  'resourceAbstractObject',
  'geom',
  'resourceIdentifier',
  'resourceType',
  'resourceTemporalExtentDetails',
  'ownerOrgForResourceObject',
  'originatorOrgForResourceObject',
  'custodianOrgForResourceObject',
  'distributorOrgForResourceObject',
  'pointOfContactOrgObject',
  'linkUrl',
  'OrgNceaClassifiers'
];

export const quickSearchTargetFields: string[] = [
  'metadataIdentifier',
  'resourceTitleObject.default',
  'resourceAltTitleObject.default',
  'resourceAbstractObject.default',
  'lineageObject.default',
  'tag.default',
  'cl_topic.default',
  'resourceType',
  'supplementalInformationObject.default',
];

export const guidedSearchSteps = {
  date: 'date',
  coordinate: 'extent',
};

export const queryParamKeys = {
  quickSearch: 'q',
  fromDateDay: 'fdd',
  fromDateMonth: 'fdm',
  fromDateYear: 'fdy',
  toDateDay: 'tdd',
  toDateMonth: 'tdm',
  toDateYear: 'tdy',
  coordinates: 'ext',
  north: 'nth',
  south: 'sth',
  east: 'est',
  west: 'wst',
  count: 'cnt',
  page: 'pg',
  rowsPerPage: 'rpp',
  sort: 'srt',
  journey: 'jry',
  resourceType: 'rty',
  startYear: 'sy',
  toYear: 'ty',
};

export const uniqueResourceTypesKey: string = 'unique_resource_types';
export const startYearRangeKey: string = 'start_year_range';
export const toYearRangeKey: string = 'to_year_range';
export const yearRange: string = 'year';

export const resourceTypeFilterField = 'resourceTypeFilter';
export const studyPeriodFilterField = 'studyPeriodFilter';

export const defaultFilterOptions: IFilterOptions = [
  {
    key: uniqueResourceTypesKey,
    needCount: true,
    propertyToRead: 'key',
    hasBucket: true,
    isTerm: true,
  },
  {
    key: yearRange,
    needCount: false,
    propertyToRead: 'value_as_string',
    hasBucket: false,
    isDate: true,
  },
];

export const detailsTabOptions: TabOptions = {
  general: {
    Abstract: 'content',
    'Study periods': 'studyPeriod',
    'Topic categories': 'topicCategories',
    Keywords: 'keywords',
    'Resource languages': 'language',
  },
  access: {
    'File Identifier': 'ncea_catalogue_number',
    'Resource Identifier': 'host_catalogue_number',
    'Coupled resources': 'host_catalogue_entry',
    'Resource type': 'resource_type_and_hierarchy',
    'Hierarchy level name': 'hierarchy_level',
    'Resource locators': 'resource_locators',
    'Catalouge entry': 'ncea_catalogue_number',
    'NCEA search service catalogue number': 'host_service_catalogue_number',
    'Parent ID': 'ncea_group_reference',
    'Metadata standard': 'metadata_standard',
    'Project number': 'project_number',
    'Metadata language': 'metadata_language',
  },
  quality: {
    'Date of publication': 'publicationInformation',
    'Date of creation': 'creationInformation',
    'Date of last revision': 'revisionInformation',
    'Metadata date': 'metadataDate',
    Lineage: 'lineage',
    Conformity: 'conformity',
    'Additional information': 'additionalInformation',
  },
  geography: {
    'Spatial data service': 'spatialDataService',
    'Spatial representation service': 'spatialRepresentationService',
    'Spatial referencing system': 'spatialReferencingSystem',
    'Geographic locations': 'geographicLocations',
    'Geographic boundary': 'geographicBoundaryHtml',
    map: 'geographicBoundary',
    'Vertical extent<br /><span>(Meters above sea level)</span>': 'verticalExtent',
    'Spatial resolution': 'samplingResolution',
  },
  license: {
    'Limitations on public access': 'limitation_on_public_access',
    'License constraints': 'license_constraints',
    'Distribution formats': 'available_formats',
    'Frequency of update': 'frequency_of_update',
    'Character encoding': 'character_encoding',
  },
};

export const pageTitles = {
  home: 'NCEA Search Service Home',
  date: 'NCEA Guided Search - Study Period',
  geography: 'NCEA Guided Search - Geographic Data',
  results: 'NCEA Search Results Summary',
  generalTab: 'NCEA Catalogue Detail - General',
  accessTab: 'NCEA Catalogue Detail - Access',
  naturalCapitalTab: 'NCEA Catalogue Detail - Natural Capital',
  qualityTab: 'NCEA Catalogue Detail - Quality',
  geographyTab: 'NCEA Catalogue Detail - Geography',
  governanceTab: 'NCEA Catalogue Detail - Governance',
  licenseTab: 'NCEA Catalogue Detail - License',
  help: 'Help',
  accessibility: 'Accessibility statement',
  termsAndConditions: 'Terms and conditions',
  privacyPolicy: 'Privacy policy',
  cookiePolicy: 'Cookies',
};
