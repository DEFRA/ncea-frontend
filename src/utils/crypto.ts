import * as crypto from 'crypto';

const { publicKey, privateKey } = crypto.generateKeyPairSync('rsa', {
  modulusLength: 2048,
  publicKeyEncoding: {
    type: 'spki',
    format: 'pem',
  },
  privateKeyEncoding: {
    type: 'pkcs8',
    format: 'pem',
  },
});

const encryptData = (data: string): string => {
  const encryptedBuffer = crypto.publicEncrypt(
    {
      key: publicKey,
      padding: crypto.constants.RSA_PKCS1_OAEP_PADDING,
      oaepHash: 'sha256',
    },
    Buffer.from(data),
  );

  return encryptedBuffer.toString('base64');
};

const decryptData = (encryptedData: string): string => {
  const decryptedBuffer = crypto.privateDecrypt(
    {
      key: privateKey,
      padding: crypto.constants.RSA_PKCS1_OAEP_PADDING,
      oaepHash: 'sha256',
    },
    Buffer.from(encryptedData, 'base64'),
  );

  return decryptedBuffer.toString();
};

export { encryptData, decryptData };
