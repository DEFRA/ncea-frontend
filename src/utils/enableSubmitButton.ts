import { IFormValidatorOptions } from '../interfaces/guidedSearch.interface';

const enableSubmitButton = async (formValidatorOptions: IFormValidatorOptions) => {
  const form = document.getElementById(formValidatorOptions.formId) as HTMLFormElement;
  const submitButton = document.getElementById(formValidatorOptions.submitButtonId) as HTMLButtonElement;
  const fields = Array.from(form.querySelectorAll('input[type="text"]')) as HTMLInputElement[];

  const checkFields = (): void => {
    const hasValue = fields.some((field) => field.value.trim() !== '');
    submitButton.disabled = !hasValue;
  };

  fields.forEach((field) => {
    field.addEventListener('input', checkFields);
  });

  checkFields();
};

const injectDynamicEnablingScript = (formValidatorOptions: IFormValidatorOptions) => {
  const script = `<script>
      (${enableSubmitButton.toString()})(${JSON.stringify(formValidatorOptions)});
    </script>`;
  return script;
};

export { enableSubmitButton, injectDynamicEnablingScript };
