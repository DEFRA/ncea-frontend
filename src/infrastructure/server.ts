'use strict';

import { environmentConfig } from '../config/environmentConfig';
import inert from '@hapi/inert';
import vision from '@hapi/vision';
import Hapi, { Server } from '@hapi/hapi';
import logger from '../infrastructure/plugins/appinsights-logger';

import { customHapiPino, customHapiRoutes, customHapiViews } from './plugins/index';

// Create the hapi server
const server: Server = Hapi.server({
  host: process.env.HOST ?? '0.0.0.0',
  port: environmentConfig.env !== 'test' ? environmentConfig.port : 4000,
  routes: {
    validate: {
      options: {
        abortEarly: false,
      },
    },
  },
});

const initializeServer = async (): Promise<Server> => {
  // Register vendors plugins
  await server.register([inert, vision]);

  // Register the custom plugins
  await server.register([customHapiViews, customHapiRoutes, customHapiPino]);

  await server.initialize();
  return server;
};

const startServer = async (): Promise<Server> => {
  logger.info('Initializing server');
  await server.start();  
  console.log(`Server running at: ${server.info.uri}`);
  return server;
};

export { initializeServer, startServer };
