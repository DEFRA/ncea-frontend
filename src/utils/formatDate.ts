import { months } from './constants';

const formatDate = (dateString: string, isFormatting: boolean = false): string => {
  const isValidDate = (dateString: string): boolean => {
    const date = new Date(dateString);
    return !isNaN(date.getTime());
  };

  if (!isValidDate(dateString)) {
    return '';
  }

  const date = new Date(dateString);
  let day: number | string = date.getDate();
  const month: string = months[date.getMonth()]!;
  const year: number = date.getFullYear();
  let hours: number | string = date.getHours();
  let minutes: number | string = date.getMinutes();
  day = day < 10 ? '0' + day : day;

  let formattedDate = isFormatting ? `${day}-${month}-${year}` : `${day} ${month} ${year}`;

  if ((hours !== 0 || minutes !== 0) && !isFormatting) {
    hours = hours < 10 ? '0' + hours : hours;
    minutes = minutes < 10 ? '0' + minutes : minutes;
    formattedDate += ` @${hours}${minutes}`;
  }

  return formattedDate;
};

export { formatDate };
