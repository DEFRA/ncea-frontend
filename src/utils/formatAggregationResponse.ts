import { getYear } from './formatDate';
import { IAggregationOption, IAggregationOptions } from '../interfaces/searchResponse.interface';
import { IFilterOption, IFilterOptions } from '../interfaces/searchPayload.interface';
import { startYearRangeKey, toYearRangeKey, uniqueResourceTypesKey } from './constants';

const capitalizeWords = (string: string): string => {
  return string.replace(/\b\w/g, (match) => match.toUpperCase());
};

const addSpaces = (string: string): string => string.split(/(?=[A-Z])/).join(' ');

type AggregationOption = {
  key: string;
};

type ClassifierValues = {
  buckets: AggregationOption[];
};

type ClassifierLevel = {
  classifier_values?: ClassifierValues;
};

type ApiResponse = {
  classifier_level?: ClassifierLevel;
};

const generateRange = (start, end): IAggregationOption[] => {
  return Array.from({ length: end - start + 1 }, (_, index) => ({
    value: String(start + index),
    text: String(start + index),
  }));
};

/* eslint-disable  @typescript-eslint/no-explicit-any */
const formatAggregationResponse = async (
  apiResponse: Record<string, any>,
  filterOptions: IFilterOptions,
): Promise<IAggregationOptions> => {
  try {
    const finalResponse: IAggregationOptions = {};

    const handleBuckets = (buckets, filterOption) => {
      let nonGeoIndex = -1;
      let publicationIndex = -1;
      buckets.forEach((bucket, index) => {
        if (bucket.key === 'nonGeographicDataset') nonGeoIndex = index;
        if (bucket.key === 'publication') publicationIndex = index;
      });

      if (nonGeoIndex !== -1 && publicationIndex !== -1) {
        buckets[nonGeoIndex].doc_count += buckets[publicationIndex].doc_count;
        buckets.splice(publicationIndex, 1);
      } else if (nonGeoIndex === -1 && publicationIndex !== -1) {
        buckets[publicationIndex].key = 'nonGeographicDataset';
      }
    };
    const formatBucket = (bucket, filterOption, capitalize = true) => {
      const wordsWithSpace = capitalize
        ? addSpaces(bucket[filterOption.propertyToRead])
        : bucket[filterOption.propertyToRead];

      let text = capitalize ? capitalizeWords(wordsWithSpace) : wordsWithSpace;
      if (filterOption.needCount) text += ` (${bucket['doc_count']})`;

      return {
        value: bucket[filterOption.propertyToRead],
        text,
      };
    };
    const handleDate = (filterOption, finalResponse) => {
      const maxYearDateString: string =
        apiResponse?.aggregations?.[`max_${filterOption.key}`]?.[filterOption.propertyToRead] ?? '';
      const minYearDateString: string =
        apiResponse?.aggregations?.[`min_${filterOption.key}`]?.[filterOption.propertyToRead] ?? '';
      if (maxYearDateString && minYearDateString) {
        const maxYear: number = Math.floor(parseInt(getYear(maxYearDateString)));
        const minYear: number = Math.floor(parseInt(getYear(minYearDateString)));
        const yearRange: IAggregationOption[] = generateRange(minYear, maxYear);
        finalResponse[startYearRangeKey] = yearRange;
        finalResponse[toYearRangeKey] = yearRange;
      } else {
        finalResponse[startYearRangeKey] = [];
        finalResponse[toYearRangeKey] = [];
      }
    };
    filterOptions.forEach((filterOption: IFilterOption) => {
      let apiAggValues = apiResponse?.aggregations?.[filterOption.key];
      const { isTerm, isDate } = filterOption;

      if (apiAggValues?.value) apiAggValues = apiAggValues.value;

      if (isTerm && apiAggValues && Array.isArray(apiAggValues.buckets) && apiAggValues.buckets.length > 0) {
        handleBuckets(apiAggValues.buckets, filterOption);

        finalResponse[filterOption.key] = apiAggValues.type === 'originator'
          ? apiAggValues.buckets.map(bucket => formatBucket(bucket, filterOption, false))
          : apiAggValues.buckets.map(bucket => formatBucket(bucket, filterOption));
      } else if (isDate) {
        handleDate(filterOption, finalResponse);
      } else {
        finalResponse[uniqueResourceTypesKey] = [];
        finalResponse[startYearRangeKey] = [];
        finalResponse[toYearRangeKey] = [];
      }
    });
    return finalResponse;
  } catch (error: any) {
    throw new Error(`Error formatting the aggregation: ${error.message}`);
  }
};

const formatClassifierResponse = async (apiResponse: ApiResponse): Promise<string[]> => {
  try {
    const finalResponse: string[] = [];
    const apiAggregationOptions = apiResponse?.classifier_level?.classifier_values?.buckets;

    if (apiAggregationOptions) {
      apiAggregationOptions.forEach((aggregationOption) => {
        finalResponse.push(aggregationOption.key);
      });
    }

    return finalResponse;
  } catch (error) {
    throw new Error(`Error formatting the aggregation: ${(error as Error).message}`);
  }
};

export { addSpaces, capitalizeWords, formatAggregationResponse, generateRange, formatClassifierResponse };
