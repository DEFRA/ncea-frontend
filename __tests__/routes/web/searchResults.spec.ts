/**
 * @jest-environment jsdom
 */

'use strict';

import { Server } from '@hapi/hapi';

import { initializeServer } from '../../../src/infrastructure/server';
import supertest from 'supertest';
import { webRoutePaths } from '../../../src/utils/constants';

jest.mock('../../../src/infrastructure/plugins/appinsights-logger', () => ({
  info: jest.fn(),
}));

jest.mock('../../../src/utils/keyvault', () => ({
  getSecret: jest.fn(),
}));

let serverRequest;

const invokeRoute = async (route, payload, method = 'post') => {
  let response;
  if (method === 'post') {
    response = await serverRequest.post(route).send(payload);
  } else {
    response = await serverRequest.get(route).query({ ...payload });
  }
  const rawHTML = response.text;
  const parser = new DOMParser();
  const document = parser.parseFromString(rawHTML, 'text/html');
  return { response, document };
};

const successPayload = {
  search_term: 'test',
  pageName: 'home',
};

describe('Results Screen', () => {
  let server: Server;

  beforeAll(async () => {
    initializeServer().then(async (s: Server) => {
      server = s;
      serverRequest = supertest(server.listener);
    });
  });

  afterAll((done) => {
    server.stop().then(() => done());
  });

  describe('Sanpshot verification & status code', () => {
    let response;

    beforeAll(async () => {
      const responseObject = await invokeRoute(
        webRoutePaths.results,
        successPayload,
      );
      response = responseObject.response;
    });

    it('should match the results screen snapshot', async () => {
      expect(response.text).toMatchSnapshot();
    });

    it('should route works with status code 302', async () => {
      expect(response.statusCode).toEqual(302);
    });
  });

  describe('Quick search POST verification on results page', () => {
    let document;
    let response;
    const redirectedRoute = `${webRoutePaths.results}?q=test&jry=qs&pg=1&rpp=20&srt=best_match&rty=all`;

    beforeAll(async () => {
      const responseObject = await invokeRoute(
        webRoutePaths.results,
        successPayload,
      );
      response = responseObject.response;
      const redirectedResponseObject = await invokeRoute(
        response.headers.location,
        {
          q: 'test',
          jry: 'qs',
          pg: '1',
          rpp: '20',
          srt: 'best_match',
          rty: 'all',
        },
      );
      document = redirectedResponseObject.document;
    });

    describe('Check the redirection', () => {
      it('should redirect to the results get route with query params', async () => {
        expect(response.status).toBe(302);
        expect(response.headers.location).toBe(redirectedRoute);
      });
    });

    describe('Breadcrumb classes', () => {
      it('renders the container class', async () => {
        expect(document.querySelector('.govuk-breadcrumbs')).toBeTruthy();
      });

      it('renders the breadcrumb list class', async () => {
        expect(document.querySelector('.govuk-breadcrumbs__list')).toBeTruthy();
      });
    });

    describe('Breadcrumb list items', () => {
      it('should render 2 list items', async () => {
        const breadcrumbList = document?.querySelector(
          '.govuk-breadcrumbs__list',
        );
        expect(breadcrumbList?.childElementCount).toEqual(2);
      });
    });

    describe('Breadcrumb list item options', () => {
      it('should render home list item first child as an anchor', async () => {
        const item = document.querySelector(
          '.govuk-breadcrumbs__list',
        )?.firstElementChild;
        const anchor = item?.firstElementChild;
        expect(anchor?.tagName.toLowerCase()).toBe('a');
        expect(anchor?.getAttribute('class')).toEqual(
          'govuk-breadcrumbs__link',
        );
        expect(anchor?.getAttribute('href')).toEqual(webRoutePaths.home);
        expect(anchor?.textContent?.trim()).toEqual('Home');
      });

      it('should not render search results list item second child as an anchor', async () => {
        const item = document.querySelector(
          '.govuk-breadcrumbs__list',
        )?.lastElementChild;
        expect(item.querySelector('a')).toBeNull();
        expect(item?.textContent?.trim()).toEqual('Search results');
      });
    });
    describe('Search block classes', () => {
      it('renders custom quick search container class', async () => {
        expect(document.querySelector('.quick_search-container')).toBeTruthy();
      });

      it('renders custom search block class', async () => {
        expect(document.querySelector('.search-block')).toBeTruthy();
      });
    });

    describe('Search block custom options', () => {
      it('renders attributes correctly', async () => {
        const searchBlock = document.querySelector('.search-block');
        expect(searchBlock?.getAttribute('data-module')).toEqual('search');
      });
    });

    describe('Search container content elements', () => {
      it('should render 2 child elements', async () => {
        const bannerContainer = document?.querySelector(
          '.quick_search-container',
        );
        expect(bannerContainer?.childElementCount).toEqual(2);
      });

      it('should not render caption on results screen', async () => {
        expect(
          document?.querySelector('.quick_search-container__caption-m'),
        ).toBeNull();
      });
    });

    describe('Search container block heading', () => {
      it('should render the search container heading', async () => {
        expect(
          document
            ?.querySelector('.quick_search-container__heading-m')
            ?.textContent?.trim(),
        ).toBe('Search');
      });

      it('should render the custom large class for heading', async () => {
        expect(
          document
            ?.querySelector('.quick_search-container__heading-m')
            ?.classList.contains('quick_search-container__heading-m--large'),
        ).toBeTruthy();
      });
    });

    describe('Search container form', () => {
      it('should render the form', async () => {
        const formElement =
          document?.querySelector('.search-block')?.firstElementChild;
        expect(formElement?.tagName.toLowerCase()).toBe('form');
      });

      it('should render the form with options', async () => {
        const formElement =
          document?.querySelector('.search-block')?.firstElementChild;
        expect(formElement?.getAttribute('role')).toBe('search');
        expect(formElement?.getAttribute('action')).toBe(webRoutePaths.results);
      });

      it('should render the form classes', async () => {
        const formElement =
          document?.querySelector('.search-block')?.firstElementChild;
        expect(
          formElement?.classList.contains('search-block__form'),
        ).toBeTruthy();
        expect(
          formElement?.firstElementChild?.classList.contains(
            'govuk-form-group',
          ),
        ).toBeTruthy();
      });
    });

    describe('Search container input field', () => {
      it('does render the input wrapper', () => {
        const wrapper = document?.querySelector(
          '.govuk-form-group > .govuk-input__wrapper',
        );
        expect(wrapper).toBeTruthy();
      });

      it('should render the input element', async () => {
        const inputElement = document?.querySelector(
          '.govuk-form-group > .govuk-input__wrapper',
        )?.firstElementChild;
        expect(inputElement?.tagName.toLowerCase()).toBe('input');
      });

      it('should renders with custom class', () => {
        const inputElement = document?.querySelector(
          '.govuk-form-group > .govuk-input__wrapper',
        )?.firstElementChild;
        expect(
          inputElement?.classList.contains('search-block__input'),
        ).toBeTruthy();
      });

      it('should renders with id', () => {
        const inputElement = document?.querySelector('.search-block__input');
        expect(inputElement?.getAttribute('id')).toEqual('search_term');
      });

      it('should renders with name', () => {
        const inputElement = document?.querySelector('.search-block__input');
        expect(inputElement?.getAttribute('name')).toEqual('search_term');
      });

      it('should renders with type="text" by default', () => {
        const inputElement = document?.querySelector('.search-block__input');
        expect(inputElement?.getAttribute('type')).toEqual('text');
      });

      it('should renders with spellcheck attribute set to false', () => {
        const inputElement = document?.querySelector('.search-block__input');
        expect(inputElement?.getAttribute('spellcheck')).toEqual('false');
      });
    });

    describe('Search container when it includes a suffix', () => {
      it('should renders the suffix inside the wrapper', () => {
        const suffix = document?.querySelector(
          '.govuk-form-group > .govuk-input__wrapper > .govuk-input__suffix',
        );
        expect(suffix).toBeTruthy();
      });

      it('should renders a button in the suffix', () => {
        const buttonElement = document?.querySelector(
          '.govuk-form-group > .govuk-input__wrapper > .govuk-input__suffix',
        )?.firstElementChild;
        expect(buttonElement?.tagName?.toLowerCase()).toBe('button');
      });

      it('should renders a button with custom class in the suffix', () => {
        const buttonElement = document?.querySelector(
          '.govuk-form-group > .govuk-input__wrapper > .govuk-input__suffix',
        )?.firstElementChild;
        expect(
          buttonElement?.classList?.contains('search-block__button'),
        ).toBeTruthy();
      });
    });
  });

  describe('Quick search POST validation on results page', () => {
    it('fail action - empty search term', async () => {
      const responseObject = await invokeRoute(webRoutePaths.results, {
        search_term: '',
        pageName: 'results',
      });
      expect(responseObject.response.statusCode).toEqual(400);
    });
    it('fail action - min length validation', async () => {
      const responseObject = await invokeRoute(webRoutePaths.results, {
        search_term: 'foo',
        pageName: 'results',
      });
      expect(responseObject.response.statusCode).toEqual(400);
    });
  });

  describe('Quick search POST validation on home page', () => {
    let document;

    beforeAll(async () => {
      const responseObject = await invokeRoute(webRoutePaths.results, {
        search_term: '',
        pageName: 'home',
      });
      document = responseObject.document;
    });

    describe('Breadcrumb classes', () => {
      it('should not render the breadcrumb class', async () => {
        expect(document.querySelector('.govuk-breadcrumbs')).toBeNull();
      });
    });

    describe('Hero content block classes', () => {
      it('renders custom full width class', async () => {
        expect(document.querySelector('.full-width-banner')).toBeTruthy();
      });

      it('renders custom banner container class', async () => {
        expect(document.querySelector('.banner-container')).toBeTruthy();
      });
    });

    describe('Hero content elements', () => {
      it('should render 2 child elements', async () => {
        const bannerContainer = document?.querySelector('.banner-container');
        expect(bannerContainer?.childElementCount).toEqual(2);
      });
    });

    describe('Hero content block heading', () => {
      it('should render the hero content heading', async () => {
        expect(
          document
            ?.querySelector('.banner-container__heading-xl')
            ?.textContent?.trim(),
        ).toBe('Natural Capital Search Service');
      });
    });

    describe('Search block classes', () => {
      it('renders custom quick search container class', async () => {
        expect(document.querySelector('.quick_search-container')).toBeTruthy();
      });

      it('renders custom search block class', async () => {
        expect(document.querySelector('.search-block')).toBeTruthy();
      });
    });

    describe('Search block custom options', () => {
      it('renders attributes correctly', async () => {
        const searchBlock = document.querySelector('.search-block');
        expect(searchBlock?.getAttribute('data-module')).toEqual('search');
      });
    });

    describe('Search container content elements', () => {
      it('should render 3 child elements', async () => {
        const bannerContainer = document?.querySelector(
          '.quick_search-container',
        );
        expect(bannerContainer?.childElementCount).toEqual(3);
      });

      it('should render caption on home screen', async () => {
        expect(
          document?.querySelector('.quick_search-container__caption-m'),
        ).toBeTruthy();
      });
    });

    describe('Search container block heading', () => {
      it('should render the search container heading', async () => {
        expect(
          document
            ?.querySelector('.quick_search-container__heading-m')
            ?.textContent?.trim(),
        ).toBe('Quick Search');
      });

      it('should not render the custom large class for heading', async () => {
        expect(
          document
            ?.querySelector('.quick_search-container__heading-m')
            ?.classList.contains('quick_search-container__heading-m--large'),
        ).toBeFalsy();
      });
    });

    describe('Search container form', () => {
      it('should render the form', async () => {
        const formElement =
          document?.querySelector('.search-block')?.firstElementChild;
        expect(formElement?.tagName.toLowerCase()).toBe('form');
      });

      it('should render the form with options', async () => {
        const formElement =
          document?.querySelector('.search-block')?.firstElementChild;
        expect(formElement?.getAttribute('role')).toBe('search');
        expect(formElement?.getAttribute('action')).toBe(webRoutePaths.results);
      });

      it('should render the form classes', async () => {
        const formElement =
          document?.querySelector('.search-block')?.firstElementChild;
        expect(
          formElement?.classList.contains('search-block__form'),
        ).toBeTruthy();
        expect(
          formElement?.firstElementChild?.classList.contains(
            'govuk-form-group',
          ),
        ).toBeTruthy();
      });
    });

    describe('Search container input field', () => {
      it('does render the input wrapper', () => {
        const wrapper = document?.querySelector(
          '.govuk-form-group > .govuk-input__wrapper',
        );
        expect(wrapper).toBeTruthy();
      });

      it('should render the input element', async () => {
        const inputElement = document?.querySelector(
          '.govuk-form-group > .govuk-input__wrapper',
        )?.firstElementChild;
        expect(inputElement?.tagName.toLowerCase()).toBe('input');
      });

      it('should renders with custom class', () => {
        const inputElement = document?.querySelector(
          '.govuk-form-group > .govuk-input__wrapper',
        )?.firstElementChild;
        expect(
          inputElement?.classList.contains('search-block__input'),
        ).toBeTruthy();
      });

      it('should renders with id', () => {
        const inputElement = document?.querySelector('.search-block__input');
        expect(inputElement?.getAttribute('id')).toEqual('search_term');
      });

      it('should renders with name', () => {
        const inputElement = document?.querySelector('.search-block__input');
        expect(inputElement?.getAttribute('name')).toEqual('search_term');
      });

      it('should renders with type="text" by default', () => {
        const inputElement = document?.querySelector('.search-block__input');
        expect(inputElement?.getAttribute('type')).toEqual('text');
      });

      it('should renders with spellcheck attribute set to false', () => {
        const inputElement = document?.querySelector('.search-block__input');
        expect(inputElement?.getAttribute('spellcheck')).toEqual('false');
      });
    });

    describe('Search container when it includes a suffix', () => {
      it('should renders the suffix inside the wrapper', () => {
        const suffix = document?.querySelector(
          '.govuk-form-group > .govuk-input__wrapper > .govuk-input__suffix',
        );
        expect(suffix).toBeTruthy();
      });

      it('should renders a button in the suffix', () => {
        const buttonElement = document?.querySelector(
          '.govuk-form-group > .govuk-input__wrapper > .govuk-input__suffix',
        )?.firstElementChild;
        expect(buttonElement?.tagName?.toLowerCase()).toBe('button');
      });

      it('should renders a button with custom class in the suffix', () => {
        const buttonElement = document?.querySelector(
          '.govuk-form-group > .govuk-input__wrapper > .govuk-input__suffix',
        )?.firstElementChild;
        expect(
          buttonElement?.classList?.contains('search-block__button'),
        ).toBeTruthy();
      });
    });

    describe('YouTube video block classes', () => {
      it('renders custom video container class', async () => {
        expect(document.querySelector('.video-container')).toBeTruthy();
      });

      it('renders custom video frame class', async () => {
        expect(document.querySelector('.video-container__frame')).toBeTruthy();
      });
    });

    describe('YouTube video block elements', () => {
      it('should render 2 child elements', async () => {
        const videoContainer = document?.querySelector('.video-container');
        expect(videoContainer?.childElementCount).toEqual(2);
      });
    });

    describe('YouTube video block heading', () => {
      it('should render the video content heading', async () => {
        expect(
          document
            ?.querySelector('.video-container__heading-m')
            ?.textContent?.trim(),
        ).toBe('What is natural capital?');
      });
    });

    describe('YouTube video frame options', () => {
      it('should render youtube video in iframe', async () => {
        const videoFrame = document?.querySelector('.video-container__frame');
        expect(videoFrame?.firstElementChild?.tagName.toLowerCase()).toBe(
          'iframe',
        );
      });

      it('should render youtube video url', async () => {
        const videoFrame = document?.querySelector('.video-container__frame');
        expect(videoFrame?.firstElementChild?.getAttribute('src')).toBe(
          'https://www.youtube.com/embed/4wx0rMruSJo?si=h-SD8VoSUSNwqo1S',
        );
      });
    });

    describe('Questionnaire search container block heading', () => {
      it('should render the questionnaire search container heading', async () => {
        expect(
          document
            ?.querySelectorAll('.govuk-heading-m')?.[1]
            ?.textContent?.trim(),
        ).toBe('Questionnaire search');
      });

      it('should not render the custom large class for heading', async () => {
        expect(
          document
            ?.querySelectorAll('.govuk-heading-m')?.[1]
            ?.classList.contains('govuk-heading-m--large'),
        ).toBeFalsy();
      });
    });

    describe('Questionnaire search block action', () => {
      it('should render the questionnaire search button', () => {
        const anchorElement = document?.querySelector('.govuk-button--start');
        expect(anchorElement?.tagName?.toLowerCase()).toBe('a');
      });
    });
  });
});
