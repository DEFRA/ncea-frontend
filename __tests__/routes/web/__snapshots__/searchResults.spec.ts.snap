// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Results Screen Sanpshot verification & status code should match the results screen snapshot 1`] = `"{"message":"An internal server error occurred","statusCode":500,"error":"Internal Server Error"}"`;
